(BLOG) Porting and Running multi node modules in cloud ci (#237)
puppetlabs/iac: _posts/2021-05-07-running-multinode-modules-gcp.md

author: sheenaajay
categories:
  - module
  - development
tags:
  - modules
  - cloud ci
  - multi node
---

## Multi Node Modules in GCP

**Introduction**

Integration testing in puppet terms, is where we set up a number of vms/containers and test interactions between them. For example this may be

* Install PE and set up some puppet agents to talk with it.
* Setup a NTP server, and register some ntp clients.
* Install puppet server and setup multiple puppet agents.

Ordering of the tests is important, and being able to run a test on an individual system is paramount.

**Design**

We use the following workflow to run multi node modules on  Cloud CI / GCP.

* Provision - We use provision task to provision a node for testing.
* Setup provisioned systems - We use bolt tasks/plans to setup the multi node environment for testing.
* Run tests - We use existing serverspec / litmus helpers to setup the dependencies and setup required for module
  testing.
* Teardown -  We use provision task to teardown the machines.

Lets go through each step in detail.

***Provision***

Litmus allows the provisioning of a vm/container via a task.

> We use the following [task](https://github.com/puppetlabs/provision/blob/main/tasks/provision_service.json) to provision machine in cloud ci.

Reviewer: ```suggestion
We use the provision module's [available tasks](https://github.com/puppetlabs/provision/blob/main/tasks) to spin up the test environment.
```

When run it creates an litmus_inventory.yaml file that allows both [bolt](https://github.com/puppetlabs/bolt) and [serverspec](https://serverspec.org/) to communicate with that vm/container.

How do we differentiate between different vm/containers? 
Provisioning tasks in the provision module allow us to add arbitrary key/pair values to the bolt inventory file. 
Using the bolt variables allows users to have multiple labels associated with a single machine.

By running the provision bolt tasks manually or through a bolt plan we can label machines.

> ![Showing role infm in litmus_inventory.yaml file]({% link /assets/2021-05-07-running-multinode-modules-gcp/inventory_role.png %}) 

Reviewer 2: ```suggestion
![Showing role info in litmus_inventory.yaml file]({% link /assets/2021-05-07-running-multinode-modules-gcp/inventory_role.png %}) 
```

For integration testing we are using bolt plans and inventory variables where we can run puppet code / commands, against the specific systems, by labelling the systems.
An example [puppetlabs-websphere_application_server](https://github.com/puppetlabs/puppetlabs-websphere_application_server/blob/main/plans/pe_server_setup.pp)

![Example for provision plan]({% link /assets/2021-05-07-running-multinode-modules-gcp/provision_plan.png %})

We can use spec/spec_helper_acceptance_local.rb for extra setup that may be required to test a module.
An example [puppetlabs-websphere_application_server](https://github.com/puppetlabs/puppetlabs-websphere_application_server/blob/main/spec/spec_helper_acceptance_local.rb)

***Run tests***

We use rspec labelling and rake task to identify the tests to run on integeration environment. Lets go through the details.

* Rspec labelling
  How do we differentiate acceptance tests from integration tests?
  Using rspec labeling we can label tests, e.g. the test is tagged as integration. Everything else can stay the same.

>   ![Example for tagging tests]({% link /assets/2021-05-07-running-multinode-modules-gcp/integration_tag.png %})

Reviewer 2: Please change screenshots of code to textual code snippets for accessibility, e.g. that people can copy/paste from there.

* Helper functions to target labeled containers/vms
  There are a number of helper methods required to allow tests to target a specific container/vm.
  This is required for serverspec/litmus as it will be carrying out the tests on that container/vm.
  Below are some examples or helper methods of filtering to get either a single vm/container or retrieving multiple vms/targets.
  [puppetlabs-websphere_application_server](https://github.com/puppetlabs/puppetlabs-websphere_application_server/blob/main/spec/spec_helper_acceptance_local.rb)
  [puppetlabs-kubernetes](https://github.com/puppetlabs/puppetlabs-kubernetes/blob/main/spec/spec_helper_acceptance_local.rb)

  ![Example for identifying target nodes]({% link /assets/2021-05-07-running-multinode-modules-gcp/target_nodes.png %})

* Testing
   The below command will only run tests labelled as ‘integration’
   `bundle exec rake websphere_application_server:integration`

***Teardown***

  We use existing provision task to teardown all the provisioned machines.

## Lets go through different scenarios where we can use multi node testing

> ***Multilpe puppet agents***

Reviewer 2: ```suggestion
***Multiple puppet agents***
```

In the setup we have many puppet agents, a module is installed and tests are run.

Commands
```
bundle install --path .bundle/gems/ --jobs 4
bundle exec rake spec_prep
bundle exec bolt --modulepath spec/fixtures/modules plan run ntp::provision_gcp
bundle exec rake litmus:install_agent
bundle exec rake litmus:install_module
bundle exec rake ntp::integration
bundle exec rake 'litmus:tear_down'
```
Example
[puppetlabs-ntp](https://github.com/puppetlabs/puppetlabs-ntp/tree/multinodentp)

***Puppet server and agent***

In the setup we have puppet server and puppet agents, a module is installed and tests are run.
we can use bolt plans for provisioning number of nodes with roles tagged to each of it.
There is new task to install open source puppet server in the provision module.
[install_puppetserver](https://github.com/puppetlabs/provision/blob/main/tasks/install_puppetserver.json)

Commands
```
bundle install --path .bundle/gems/ --jobs 4
bundle exec rake spec_prep
bundle exec bolt --modulepath spec/fixtures/modules plan run kubernetes::provision_cluster

> bundle exec bolt --modulepath spec/fixtures/modules plan run kubernetes::puppetserver_setup

Reviewer 2: don't these also need `--inventory spec/fixtures/litmus_inventory.yaml` ?

Author: absolutely. Adding the changes . Thank you
